### ISSUE TITLE
NFL Tour - Wrong/Missing Colors

### ISSUE THREAD

**Issue by ChwaraeTeg** [RPCS3/rpcs3#12262]
## Quick summary
NFL Tour has some graphical issues with incorrect colors

## Details
NFL Tour has some graphical issues with RPCS3 showing wrong/missing colors on the field as compared to real hardware. The grass appears yellow instead of green; distance line markings, logos etc. appear black instead of the expected colors; end zone is not the right shade of blue. 

#### 0. Make sure you're running with settings as close to default as possible
Game requires LLE libvdec to get in-game. Have tried leaving all other settings default, setting Write Color Buffers, setting WCB+RCB+WDB+RDB, and setting all 4 of those + Force CPU Blit - none of these made any difference.

#### 1. Please attach RPCS3's log.
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/8943566/RPCS3.log.gz)

#### 2. If you describe graphical issue, please provide an RSX capture and a RenderDoc capture that demonstrate it.
RSX Capture:
https://mega.nz/file/GN0TkJ4b#ka0nXknagLlabrsnDSji27mejCwEnPwrgLFi8qR0BS8

#### 3. Please attach screenshots of your problem.
![RPCS3](https://user-images.githubusercontent.com/104720038/174676209-bc61a53e-be5d-4f9d-b1a4-228843371f62.jpg)

#### 4. Please provide comparison with real PS3.
![NFLTourPS3](https://user-images.githubusercontent.com/104720038/174676226-f0d96dd3-454f-4ca9-859f-b31256a4f583.png)

#### 5. Please provide your system configuration:
OS - Windows 10 21H2
CPU Intel Xeon E3-1240 v5
GPU NVIDIA GeForce GTX 1060 6GB
Driver version 512.95.0.0

#### Please include.
* Anything else you deem to be important

**Comment by kd-11** [comment]
Still an issue?

**Comment by Triticum0** [comment]
Yes

**Comment by ChwaraeTeg** [comment]
RenderDoc capture if it helps:
https://mega.nz/file/7AkXXZxS#k6T-LAQTS8noXTY4Ux9Wfi_YKpcR-lO4DzsXAaotm5M

Corresponding log: 
[nfltour.log.gz](https://github.com/RPCS3/rpcs3/files/11502342/nfltour.log.gz)

**Comment by Triticum0** [comment]
Retested with v0.0.30-15916-80b7fd5f. No changes

**Comment by kd-11** [comment]
This one looks interesting.

**Comment by kd-11** [comment]
This one is insanity.
The game basically disables an attribute channel then proceeds to read from it in shaders.
```C
enable_attr(0, ...);
disable_attr(3, ...);
disable_attr(8, ...);

// Some other setup

enable_attr(0, ...);
enable_attr(8, ...);

draw();
```

The shader expects attr 3 to be present (diffuse color intensity). Without being set we fall back to default (verified with hardware tests years ago) which is alpha value only of 1.
There may be a method to this madness however, if they actually just want to pick the alpha channel. But why not shuffle it to "on" state in the shader by just picking .a instead?
The color target is of type "B8" - I wonder if the rendering is configured to pick from "A"?

**Comment by kd-11** [comment]
Truncating the word so that A is mapped to B gives a completely overblown image, the pitch colors are correct but way too bright. I feel we're missing something else, this one needs some hardware tests.

**Comment by mysteria25** [comment]
@kd-11   This game used to show proper colors but regressed it seems.

Build 0.0.6-8647- Field colors / lighting work correctly except the field is very grainy no matter what resolution or AF you use. https://github.com/RPCS3/rpcs3-binaries-win/releases/download/build-14aa3b3360e22ec78427202aaf37ff1e17d53999/rpcs3-v0.0.7-8647-14aa3b33_win64.7z


https://github.com/user-attachments/assets/44385532-9974-42b2-acdd-92a7d1c009f7





Build 0.0.7-8650 - Field Colors / lighting regresses and has been that way ever since afaik. 
https://github.com/RPCS3/rpcs3-binaries-win/releases/download/build-efa501dac67bc8980f9c85036dca5d417a90ad9c/rpcs3-v0.0.7-8650-efa501da_win64.7z


https://github.com/user-attachments/assets/7d9876bb-432e-48fe-8ac7-feea0092e85b



Regression?   PR = #6485

Settings: Vulkan, WCB + Force CPU Blit.   Firmware: Automatic + libvdec enabled.  ASMJIT SPU Recompiler
Windows 10

**Comment by kd-11** [comment]
Thanks. This narrows things down a bit.

**Comment by kd-11** [comment]
As discovered earlier it has to do with the disabled input attribute.
I need to do extensive testing on that as well.

**Comment by digant73** [comment]
@mysteria25, @ChwaraeTeg is that still an issue?

**Comment by Triticum0** [comment]
Yes

<img width="1280" height="720" alt="Image" src="https://github.com/user-attachments/assets/07169595-1bf6-413c-a2dc-16ab7bc5c9aa" />

**Comment by digant73** [comment]
this bug is fixed simply restoring the following code in function `vec4 read_location(const in int location)` (now present in `RSXVertexFetch.glsl`) that was removed in #6485:

```
	if (desc.attribute_size == 0)
	{
		//default values
		const vec4 defaults[] =
		{
			vec4(0., 0., 0., 1.), //position
			vec4(0.), vec4(0.), //weight, normals
			vec4(1.), //diffuse
			vec4(0.), vec4(0.), //specular, fog
			vec4(1.), //point size
			vec4(0.), //in_7
			//in_tc registers
			vec4(0.), vec4(0.), vec4(0.), vec4(0.),
			vec4(0.), vec4(0.), vec4(0.), vec4(0.)
		};
		return defaults[location];
	}
```

**Comment by kd-11** [comment]
> this bug is fixed simply restoring the following code in function `vec4 read_location(const in int location)` (now present in `RSXVertexFetch.glsl`) that was removed in [#6485](https://github.com/RPCS3/rpcs3/pull/6485):
> 
> ```
> 	if (desc.attribute_size == 0)
> 	{
> 		//default values
> 		const vec4 defaults[] =
> 		{
> 			vec4(0., 0., 0., 1.), //position
> 			vec4(0.), vec4(0.), //weight, normals
> 			vec4(1.), //diffuse
> 			vec4(0.), vec4(0.), //specular, fog
> 			vec4(1.), //point size
> 			vec4(0.), //in_7
> 			//in_tc registers
> 			vec4(0.), vec4(0.), vec4(0.), vec4(0.),
> 			vec4(0.), vec4(0.), vec4(0.), vec4(0.)
> 		};
> 		return defaults[location];
> 	}
> ```

It was changed because running a test on a real PS3 showed it was wrong. It's not getting reverted to emit made up constants unless the console says otherwise. As of now, it still says output is (0, 0, 0, 1) so something else is broken and this is just the unhider.

**Comment by digant73** [comment]
Yes I read it. I simply reverted the piece of code apparently fixing the issue. It also seems that the `location` value fixing the issue was the one related to:

`vec4(1.), //diffuse`

It seems also that the only coordinate of the vec4 needed to fix the issue was the first one. So with something like:

`vec4(1., 0.,0.,0.), //diffuse`

the issue is still not present.

Lowering the first coordinate's value from 1. to 0. will gradually convert the field's green color to a dark yellow color

**Comment by kd-11** [comment]
We already knew all that, attribute 3 is explicitly disabled by the game then the shader reads from it (see old comments). In that scenario hardware always returns (0, 0, 0, 1) so the problem happens some time before the shader is even called.

**Comment by audiowylde2345** [comment]
Is this still an issue

**Comment by digant73** [comment]
yes

**Comment by digant73** [comment]
@kd-11 please let me know in case the 2 fixes deserve a PR for a review

in case it makes sense, the following reset on method `void rsx_state::init()` fixes the regression, The game however needs anisotropic filter 16x to reduce a sort of distortion at distance (not present on console):

```
		// Reset state of the constant vertex attribute register file.
		// When a vertex program reads an attribute that has neither an array bound nor a constant register
		// written by the game, the hardware returns the reset value of the corresponding register.
		// The generic reset value is (0, 0, 0, 1) - this matches both the NV generic attribute default and
		// the GL fixed-function defaults for normals, texture coordinates (q = 1) and secondary color.
		// The two exceptions are the attributes GL defines as one: the primary color and the point size.
		// NOTE: Values are stored in PS3-native (BE) order because the register upload path in
		// draw_command_processor::fill_vertex_layout_state always sets the byte swap flag.
		{
			static constexpr f32 attrib_reset_values[16][4] =
			{
				{ 0.f, 0.f, 0.f, 1.f }, // 0  position
				{ 0.f, 0.f, 0.f, 1.f }, // 1  weight
				{ 0.f, 0.f, 0.f, 1.f }, // 2  normal
				{ 1.f, 1.f, 1.f, 1.f }, // 3  diffuse
				{ 0.f, 0.f, 0.f, 1.f }, // 4  specular
				{ 0.f, 0.f, 0.f, 1.f }, // 5  fog
				{ 1.f, 1.f, 1.f, 1.f }, // 6  point size
				{ 0.f, 0.f, 0.f, 1.f }, // 7
				{ 0.f, 0.f, 0.f, 1.f }, // 8  texcoord0
				{ 0.f, 0.f, 0.f, 1.f }, // 9  texcoord1
				{ 0.f, 0.f, 0.f, 1.f }, // 10 texcoord2
				{ 0.f, 0.f, 0.f, 1.f }, // 11 texcoord3
				{ 0.f, 0.f, 0.f, 1.f }, // 12 texcoord4
				{ 0.f, 0.f, 0.f, 1.f }, // 13 texcoord5
				{ 0.f, 0.f, 0.f, 1.f }, // 14 texcoord6
				{ 0.f, 0.f, 0.f, 1.f }  // 15 texcoord7
			};

			for (u32 i = 0; i < 16; i++)
			{
				auto& reg = register_vertex_info[i];
				reg.type = rsx::vertex_base_type::f;
				reg.size = 4;
				reg.frequency = 0;
				reg.stride = 0;

				for (u32 c = 0; c < 4; c++)
				{
					reg.data[c] = std::bit_cast<u32, be_t<u32>>(std::bit_cast<u32>(attrib_reset_values[i][c]));
				}
			}
		}
```

EDIT: The issue on wrong pitch was identified on mip level and fixed by a small change on `texture_cache.h`.

### Fix 1: colors fixed but pitch not properly rendered (see the pitch at distance)

<img width="3838" height="1278" alt="Image" src="https://github.com/user-attachments/assets/f147774b-743f-4c89-9c47-7996a137f0e0" />

### Fix 2: pitch properly rendered (one-to-one match with console)

<img width="3838" height="2070" alt="Image" src="https://github.com/user-attachments/assets/c822715a-a0eb-4637-b725-874dae1db0b6" />

**Comment by audiowylde2345** [comment]
> in case it makes sense, the following reset on method `void rsx_state::init()` fixes the regression, The game however needs anisotropic filter 16x to reduce a sort of distortion at distance (not present on console):
> 
> ```
> 		// Reset state of the constant vertex attribute register file.
> 		// When a vertex program reads an attribute that has neither an array bound nor a constant register
> 		// written by the game, the hardware returns the reset value of the corresponding register.
> 		// The generic reset value is (0, 0, 0, 1) - this matches both the NV generic attribute default and
> 		// the GL fixed-function defaults for normals, texture coordinates (q = 1) and secondary color.
> 		// The two exceptions are the attributes GL defines as one: the primary color and the point size.
> 		// NOTE: Values are stored in PS3-native (BE) order because the register upload path in
> 		// draw_command_processor::fill_vertex_layout_state always sets the byte swap flag.
> 		{
> 			static constexpr f32 attrib_reset_values[16][4] =
> 			{
> 				{ 0.f, 0.f, 0.f, 1.f }, // 0  position
> 				{ 0.f, 0.f, 0.f, 1.f }, // 1  weight
> 				{ 0.f, 0.f, 0.f, 1.f }, // 2  normal
> 				{ 1.f, 1.f, 1.f, 1.f }, // 3  diffuse
> 				{ 0.f, 0.f, 0.f, 1.f }, // 4  specular
> 				{ 0.f, 0.f, 0.f, 1.f }, // 5  fog
> 				{ 1.f, 1.f, 1.f, 1.f }, // 6  point size
> 				{ 0.f, 0.f, 0.f, 1.f }, // 7
> 				{ 0.f, 0.f, 0.f, 1.f }, // 8  texcoord0
> 				{ 0.f, 0.f, 0.f, 1.f }, // 9  texcoord1
> 				{ 0.f, 0.f, 0.f, 1.f }, // 10 texcoord2
> 				{ 0.f, 0.f, 0.f, 1.f }, // 11 texcoord3
> 				{ 0.f, 0.f, 0.f, 1.f }, // 12 texcoord4
> 				{ 0.f, 0.f, 0.f, 1.f }, // 13 texcoord5
> 				{ 0.f, 0.f, 0.f, 1.f }, // 14 texcoord6
> 				{ 0.f, 0.f, 0.f, 1.f }  // 15 texcoord7
> 			};
> 
> 			for (u32 i = 0; i < 16; i++)
> 			{
> 				auto& reg = register_vertex_info[i];
> 				reg.type = rsx::vertex_base_type::f;
> 				reg.size = 4;
> 				reg.frequency = 0;
> 				reg.stride = 0;
> 
> 				for (u32 c = 0; c < 4; c++)
> 				{
> 					reg.data[c] = std::bit_cast<u32, be_t<u32>>(std::bit_cast<u32>(attrib_reset_values[i][c]));
> 				}
> 			}
> 		}
> ```
> 
> ### anisotropic filter on Auto (see the pitch at distance)
> <img alt="Image" width="2000" height="666" src="https://private-user-images.githubusercontent.com/64427768/632625201-f147774b-743f-4c89-9c47-7996a137f0e0.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODYwNTU3OTksIm5iZiI6MTc4NjA1NTQ5OSwicGF0aCI6Ii82NDQyNzc2OC82MzI2MjUyMDEtZjE0Nzc3NGItNzQzZi00Yzg5LTljNDctNzk5NmExMzdmMGUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA4MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwODA2VDIyMzEzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg3Nzk2OWNiNzA5MmJjOTcyMDhiMTFiNjg4NTIzZjZkOWJhOTNjYmQ4M2E5NDhiNTQ2NzY3NGQyN2I3MWI0ODAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.6rvSEc3Qu7PUbSOgndsfeHmXOYX64JeZHlpcNMlTJ4o">

I apologize if this is in any way a stupid question but is this something that is fixed in current builds and I just have to put filtering to 16x or is this something that you manually have to do every time? I'm somewhat tech savvy but I'm not remotely a program and I don't understand a lot of the workings of these emulators again I apologize if this is a stupid question.

**Comment by digant73** [comment]
no, it is only a test so not present on current official release

**Comment by audiowylde2345** [comment]
> no, it is only a test so not present on current official release

You have to do this every time you want to play this game and then you have to change it back? Or is it something I can set once and then leave it alone again sorry the dumb questions

**Comment by digant73** [comment]
> We already knew all that, attribute 3 is explicitly disabled by the game then the shader reads from it (see old comments). In that scenario hardware always returns (0, 0, 0, 1) so the problem happens some time before the shader is even called.

@kd-11 see my previous post https://github.com/RPCS3/rpcs3/issues/12262#issuecomment-5209461551

**Comment by kd-11** [comment]
> [@kd-11](https://github.com/kd-11) please let me know in case the 2 fixes deserve a PR for a review

It's the same hack as before just dressed up with many lines of code to seem legit.
You can get the same effect in the vertex shader code by just changing the default value for diffuse to be white instead of opaque black - a one-line change. But hardware testing has thoroughly disproven that behavior already, it was one of the very first things I ever checked with real hardware.

**Comment by digant73** [comment]
> > [@kd-11](https://github.com/kd-11) please let me know in case the 2 fixes deserve a PR for a review
> 
> It's the same hack as before just dressed up with many lines of code to seem legit. You can get the same effect in the vertex shader code by just changing the default value for diffuse to be white instead of opaque black - a one-line change. But hardware testing has thoroughly disproven that behavior already, it was one of the very first things I ever checked with real hardware.

ok. Even using the hack (this one or your old code) the game is still bugged (bug 2). In the following link you can see also the second possible fix https://github.com/RPCS3/rpcs3/commit/e9a7d07f3de89e4fd6b8d167d0073cbd3234061f

**Comment by kd-11** [comment]
> > > [@kd-11](https://github.com/kd-11) please let me know in case the 2 fixes deserve a PR for a review
> > 
> > 
> > It's the same hack as before just dressed up with many lines of code to seem legit. You can get the same effect in the vertex shader code by just changing the default value for diffuse to be white instead of opaque black - a one-line change. But hardware testing has thoroughly disproven that behavior already, it was one of the very first things I ever checked with real hardware.
> 
> ok. Even using the hack the game is still bugged (bug 2). In the following link you can see also the second possible fix [e9a7d07](https://github.com/RPCS3/rpcs3/commit/e9a7d07f3de89e4fd6b8d167d0073cbd3234061f)

That's not a proper fix. I was handling that in the recent PR that reuploads everything and builds the list on top, but I have paused due to edge cases uncovered with aliasing formats and upscaling.

**Comment by digant73** [comment]
> > > > [@kd-11](https://github.com/kd-11) please let me know in case the 2 fixes deserve a PR for a review
> > > 
> > > 
> > > It's the same hack as before just dressed up with many lines of code to seem legit. You can get the same effect in the vertex shader code by just changing the default value for diffuse to be white instead of opaque black - a one-line change. But hardware testing has thoroughly disproven that behavior already, it was one of the very first things I ever checked with real hardware.
> > 
> > 
> > ok. Even using the hack the game is still bugged (bug 2). In the following link you can see also the second possible fix [e9a7d07](https://github.com/RPCS3/rpcs3/commit/e9a7d07f3de89e4fd6b8d167d0073cbd3234061f)
> 
> That's not a proper fix. I was handling that in the recent PR that reuploads everything and builds the list on top, but I have paused due to edge cases uncovered with aliasing formats and upscaling.

hmm if you refer to [kd-11:2026/surface-cache-unification](https://github.com/kd-11/rpcs3/tree/2026/surface-cache-unification) then it is not working (I simply applied the hack on top of it)

<img width="3818" height="1354" alt="Image" src="https://github.com/user-attachments/assets/2de16221-6476-4cee-9ee9-42b93fd85100" />

**Comment by kd-11** [comment]
> hmm if you refer to [kd-11:2026/surface-cache-unification](https://github.com/kd-11/rpcs3/tree/2026/surface-cache-unification) then it is not working (I simply applied the hack on top of it)

No, not that PR. The previously merged one. It is not complete but that approach is the right fix.

**Comment by audiowylde2345** [comment]
Is this on an official release  or is it still being te sted
